### PR TITLE
Feature/dtls encryption

### LINKED ISSUE #1479: [SDL 0040] DTLS encryption
Proposal: [DTLS encryption](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0040-DTLS-encryption.md)

> This proposal is to add DTLS (Datagram Transport Layer Security) to SDL core. SDL core will utilize the DTLS communications protocol as a preferred solution when sending encrypted data over protected services (such audio 0x0a and video 0x0b ) including the handshake, encryption and decryption processes.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/126

Steering Committee Decision:

> The Steering Committee has agreed to accept this proposal. They've also agreed that we should target Android and iOS library changes as an additional proposal, as this cannot be tested without the library portion.

### PR DESCRIPTION
**Pull request by AByzhynar**
This PR is **ready** for review.

Fixes #1479 

### Risk
This PR makes **no** API changes.

### Testing Plan
Unit-tests are provided

### Summary
Implemented DTLS support

**NOTE**
This feature should be merged **after** #2105 and **before** security fixes #2217 + #2218 

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by LuxoftAKutsan** [comment]
Please specify github issue

**Comment by LuxoftAKutsan** [comment]
@AByzhynar please check unit tests

**Comment by jacobkeeler** [comment]
Duplicate of #1489?

**Comment by AByzhynar** [comment]
> @jacobkeeler Duplicate of #1489 

@jacobkeeler  Please read my answer below. 
This is extended implementation. 
In #1489 there are missed the following use cases: 
 
- if the same services will be specified by mistake or any other reason (in smartdevicelink.ini configuration file) as ForceProtectedService and ForceUnprotectedService - SDL stops its work as this mismatch is critical. This is wrong security configuration.
 
- If mobile app wants to send huge message which can not be sent at once due to MTU restrictions (MTU size depends on protocol version used) : 
App sends this message by frames (according to FORD protocol specification). In case of mobile app wants to send such message secured (encrypted) -  
SDL must be able to decrypt such message(every frame must be decrypted independently on previous ones) which was implemented in #2068. 
#1489  can not decrypt all frames independently. It requires that first frame must be non-secured and all consecutive frames can be secured.  
This idea breaks the logic of security encryption. What is the sense to mix encrypted and decrypted frames of the same message? 
The entire message must be SECURED or NON-SECURED.  
Current feature implementation (PR #2068 ) allows to use mixed secured & non-secured frames as well as fully secured frames.

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 1]
```diff
@@ -1140,6 +1140,10 @@ bool PolicyManagerImpl::InitPT(const std::string& file_name,
   if (ret) {
     RefreshRetrySequence();
     update_status_manager_.OnPolicyInit(cache_->UpdateRequired());
+    const std::string certificate_data = cache_->GetCertificate();
```
Comment (mrapitis) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc]: Should these changes be included in issue #2190 or #2191 instead?  Those two PR's are targeting certificate processing.

File: src/components/security_manager/src/security_manager_impl.cc [hunk 2]
```diff
@@ -143,7 +143,7 @@ security_manager::SSLContext* SecurityManagerImpl::CreateSSLContext(
 
   security_manager::SSLContext* ssl_context = session_observer_->GetSSLContext(
       connection_key, protocol_handler::kControl);
-  // return exists SSLCOntext for current connection/session
+  // return existed SSLCOntext for current connection/session
```
Comment (mrapitis) [inline src/components/security_manager/src/security_manager_impl.cc]: should be // return existing SLContext for current connection/session

File: src/components/policy/policy_regular/src/policy_manager_impl.cc (hunk 1 above)
Comment (AKalinich-Luxoft) [reply src/components/policy/policy_regular/src/policy_manager_impl.cc]: @mrapitis 
This functionality was moved out from this feature branch in https://github.com/smartdevicelink/sdl_core/pull/2068/commits/f52af91122b5daf719747027a9fa248419dc6850
Will be included to some of bug fixes pull requests

File: src/components/security_manager/src/security_manager_impl.cc (hunk 2 above)
Comment (AKalinich-Luxoft) [reply src/components/security_manager/src/security_manager_impl.cc]: @mrapitis thanks, fixed in https://github.com/smartdevicelink/sdl_core/pull/2068/commits/f52af91122b5daf719747027a9fa248419dc6850

**Comment by mrapitis** [comment]
@theresalech this PR is ready for review.

**Comment by AKalinich-Luxoft** [comment]
Rebuild required

**Pull request merged**